### ISSUE TITLE
Use new provider version or config during refresh

### ISSUE THREAD

**Issue by danielrbradley** [pulumi/pulumi#13860]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

When performing major (and sometimes minor) version upgrades of providers it would be helpful to be able to perform a read (via refresh) using the new version of the provider before attempting a deployment. Some of these breaking changes might affect properties which cause replacement, meaning that if the state and the updated program don't match, Pulumi will always try to re-create the resource even when there's no actual change required.

Currently refresh doesn't interact with the Pulumi program and will only use the version of the provider which created the state. This makes it impossible to use the new version of the provider to re-read the state so that it matches with the user's updated program.

### Specify Provider Version via the CLI

The simplest approach could be to include a CLI option to specify the version of providers to use during a refresh operation.

For example:
```
pulumi refresh --providerVersion "azure-native=2.0.1"
```

### Upgrade via a Program Update

An alternative approach is for the user to use the program to define the provider version to use.

1. The user upgrades their SDK to the new version.
2. When running `pulumi refresh`, we will run their program only to retrieve the new version of the provider they're using.
3. The engine uses the version of the provider associated with their program to perform the update.

This could perhaps be opt-in behaviour with a CLI argument such `--with-program` or a similar config variable. This could also potentially be the default behaviour in future versions.

**Comment by Frassle** [comment]
Related https://github.com/pulumi/pulumi/issues/9878

**Comment by t0yv0** [comment]
Can a provider itself decide to run a refresh (on vNext) when doing an Update on an old state? Or there's not enough information passed to the provider to detect it's in an upgrade context?

Bridged providers have notion of state migration so they are aware when resources states were provisioned on the old version, and they automatically "migrate" or translate these states to new form during normal pulumi up operation. This mechanism is supposed to cover for the scenarios of changing the representation to avoid breaks.

**Comment by Frassle** [comment]
> Or there's not enough information passed to the provider to detect it's in an upgrade context?

Not unless the providers recorded something, thus why Dan also raised https://github.com/pulumi/pulumi/issues/13861.

> Bridged providers have notion of state migration

We should probably find a way to make that clean and easy for all providers to do.

**Comment by iwahbe** [comment]
This is related to https://github.com/pulumi/pulumi-aws/issues/3149. In that case, the provider needs to read a new `config` for refresh, but since refresh uses the provider in state, this is currently impossible.

**Comment by fitz-vivodyne** [comment]
>This is related to https://github.com/pulumi/pulumi-aws/issues/3149. In that case, the provider needs to read a new config for refresh, but since refresh uses the provider in state, this is currently impossible.

More context to this.

It appears that myself and the OP  from https://github.com/pulumi/pulumi-aws/issues/3149 are attempting to use different roles/configurations for the `preview/refresh` and `up` phases to enabled least privileged access  for pull request reviews etc (we're ok with a PR having read only/refresh permissions, but only protected branches should be able to apply).

The way this currently works forces you to use the production provider config from the state file, ignoring the config you're explicitly passing it to use, breaking this capability

**Comment by danielrbradley** [comment]
Could we enable this option just while running `pulumi up --refresh`? cc: @Frassle 

If we're going to be running the program already for the `up`, could we read in the provider version before the prior refresh?

**Comment by corymhall** [comment]
I think this should now be possible with `--run-program`

**Comment by lunaris** [comment]
Is this solved now that #19578 is merged/released?

**Comment by corymhall** [comment]
Yes I believe it is. I've tested this as part of the aws v7 work and when running `pulumi up --refresh --run-program` Read is called on the new v7 provider.

**Comment by lunaris** [comment]
Nice! Will close as fixed 🙏
